src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vulkan_commands.h

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 The gf-layers Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VK_LAYER_AMBER_SCOOP_VULKAN_COMMANDS_H
+#define VK_LAYER_AMBER_SCOOP_VULKAN_COMMANDS_H
+
+#include <vulkan/vulkan.h>
+
+#include <cstdint>
+
+#include "vk_layer_amber_scoop/vk_deep_copy.h"
+
+namespace gf_layers::amber_scoop_layer {
+
+class CmdBeginRenderPass;
+class CmdBindPipeline;
+class CmdDraw;
+class CmdDrawIndexed;
+
+class Cmd {
+ public:
+  enum Kind {
+    kBeginRenderPass,
+    kBindPipeline,
+    kDraw,
+    kDrawIndexed,
+  };
+
+  explicit Cmd(Kind kind) : kind_(kind) {}
+
+  virtual ~Cmd();
+
+  // A bunch of methods for casting this type to a given type. Returns this if
+  // the cast can be done, nullptr otherwise.
+  // clang-format off
+#define DeclareCastMethod(target)                                              \
+  virtual Cmd##target *As##target() { return nullptr; }                        \
+  virtual const Cmd##target *As##target() const { return nullptr; }
+  DeclareCastMethod(BeginRenderPass)
+  DeclareCastMethod(BindPipeline)
+  DeclareCastMethod(Draw)
+  DeclareCastMethod(DrawIndexed)
+#undef DeclareCastMethod
+
+  Kind kind_;
+  // clang-format on
+
+  // Disabled copy/move constructors and copy/move assign operators
+  Cmd(const Cmd&) = delete;
+  Cmd(Cmd&&) = delete;
+  Cmd& operator=(const Cmd&) = delete;
+  Cmd& operator=(Cmd&&) = delete;
+};
+
+class CmdBeginRenderPass : public Cmd {
+ public:
+  CmdBeginRenderPass(VkRenderPassBeginInfo const* pRenderPassBegin,
+                     VkSubpassContents contents)
+      : Cmd(kBeginRenderPass),
+        render_pass_begin_(DeepCopy(pRenderPassBegin)),
+        contents_(contents) {}
+
+  CmdBeginRenderPass* AsBeginRenderPass() override;
+  [[nodiscard]] const CmdBeginRenderPass* AsBeginRenderPass() const override;
+
+  ~CmdBeginRenderPass() override { DeepDelete(&render_pass_begin_); }
+
+  VkRenderPassBeginInfo render_pass_begin_;
+  VkSubpassContents contents_;
+
+  // Disabled copy/move constructors and copy/move assign operators
+  CmdBeginRenderPass(const CmdBeginRenderPass&) = delete;
+  CmdBeginRenderPass(CmdBeginRenderPass&&) = delete;
+  CmdBeginRenderPass& operator=(const CmdBeginRenderPass&) = delete;
+  CmdBeginRenderPass& operator=(CmdBeginRenderPass&&) = delete;
```

Comment:
Looking at how these are used, I would be OK with making them structs. They are basically just deep copies of other structs anyway, and once you get access to one of the structs, you could start mutating it, so there is not much benefit from "hiding" our fields.

Reply:
I agree that these should be structs. One thing about field/parameter naming: if we use structs, then the field names shouldn't have the trailing underscore (https://google.github.io/styleguide/cppguide.html#Variable_Names) and that causes variable shadow warning in constructor initializer list. F.ex. `explicit Cmd(Kind kind) : kind(kind) {}` is accepted by the language (and works as expected), but causes warning `constructor parameter 'kind' shadows the field 'kind' of 'Cmd' [-Wshadow-field-in-constructor]`

So, should we keep the trailing underscore in the member variables to avoid this issue?
Also, looks like I forgot to change the parameter names to lowercase with underscore in this file.

Reply:
I would say don't keep the trailing underscore and then either:

 - Put a "_param" suffix on the constructor parameters, or
 - Use a pragma across the entire file.

```
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wshadow-field-in-constructor"
...
#pragma clang diagnostic pop
```

Reply:
Oh, well. I changed them to structs, but looks like clang-tidy treats structs as classes if there's any functions defined and therefore it will give a warning about public members.